feat(web): plugin playground/upload file and download plugin

devshun opened the pull request:
# Overview

## What I've done

- Function to add plugin files
- Function to rename plugin files
- Function to delete plugin files
- Function to upload plugin files
- Function to download plugin files

## What I haven't done

## How I tested

## Which point I want you to review particularly

## Memo


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **New Features**
  - Introduced a `Code` component for displaying source code with syntax highlighting.
  - Added a `FileListItem` component for managing file interactions, including renaming and deleting.
  - Enhanced `PluginInspector` to allow downloading plugins with a new button.
  - Integrated a custom hook for managing plugins and files, improving state management.

- **Improvements**
  - Updated `Plugins` component to utilize dynamic data instead of static mock data.
  - Enhanced validation for file titles to ensure uniqueness and correct file types.

These changes enhance user interaction and improve the overall functionality of the Plugin Playground feature.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

devshun commented on web/src/beta/features/PluginPlayground/Plugins/constants.ts:
```diff
@@ -0,0 +1,31 @@
+  disableDelete: true
+} as const satisfies FileType;
+
+export const ALLOWED_FILE_EXTENSIONS = ["js"] as const;
```
Since reearth.yml exists by default, newly added files are limited to .js only.

devshun commented on web/src/beta/features/PluginPlayground/Plugins/constants.ts:
```diff
@@ -0,0 +1,31 @@
+                type: widget
+                name: Test
+    `,
+  disableEdit: true,
+  disableDelete: true
```
Editing or deleting reearth.yml is restricted.

devshun commented on web/src/beta/features/PluginPlayground/Plugins/utils.ts:
```diff
@@ -0,0 +1,38 @@
+import { ALLOWED_FILE_EXTENSIONS } from "./constants";
+
+export const validateFileTitle = (
+  fileTitle: string,
+  existingFileTitles: string[]
+) => {
+  if (fileTitle === "") {
+    return {
+      success: false,
+      message: "File name cannot be empty"
+    } as const;
+  }
+  if (existingFileTitles.includes(fileTitle)) {
+    return {
+      success: false,
+      message: "File name already exists"
+    } as const;
+  }
+
+  const extension = fileTitle.split(".").pop();
+  if (
+    extension === undefined ||
+    !ALLOWED_FILE_EXTENSIONS.includes(
+      extension as Readonly<typeof ALLOWED_FILE_EXTENSIONS>[number]
+    )
+  ) {
+    return {
+      success: false,
+      message: "Invalid file type"
+    } as const;
+  }
+
+  return {
+    success: true
+  } as const;
+};
```
I have implemented validation for file titles. If validation fails during file addition, renaming, or uploading, the changes will not be reflected on the screen.

mulengawilfred commented on web/src/beta/features/PluginPlayground/Plugins/hook.ts:
```diff
@@ -0,0 +1,188 @@
+
+      if (!result.success) {
+        return;
+      }
```
We can make sure of the `useNotification` hook to create notifications when a file title is not value or any process is unsuccessful so that the user is informed of the error

mulengawilfred commented on web/src/beta/features/PluginPlayground/Plugins/hook.ts:
```diff
@@ -0,0 +1,188 @@
+    const reader = new FileReader();
+
+    reader.onload = async (e2) => {
+      const body = e2?.target?.result;
```
is it okay if we use a more readable prop name like `event` or any other that might give more description to the prop?

mulengawilfred commented on web/src/beta/features/PluginPlayground/Plugins/hook.ts:
```diff
@@ -0,0 +1,188 @@
+      const body = e2?.target?.result;
+      if (typeof body != "string") return;
+      const fileItem = {
+        id: generateUniqueId(),
```
lets use `uuid` package for unique ids

devshun replied:
fixed same issue 
https://github.com/reearth/reearth-visualizer/pull/1216#discussion_r1832646557

mulengawilfred commented on web/src/beta/features/PluginPlayground/Plugins/hook.ts:
```diff
@@ -0,0 +1,188 @@
+    reader.readAsText(file);
+  });
+
+  const handlePluginDownload = useCallback(async () => {
+    const zip = new JSZip();
+    const pluginFolder = zip.folder(selectedPlugin.title);
+
+    selectedPlugin.files.forEach((file) => {
+      pluginFolder?.file(file.title, file.sourceCode);
+    });
+
+    const zipBlob = await zip.generateAsync({ type: "blob" });
+    const zipUrl = URL.createObjectURL(zipBlob);
+    const link = document.createElement("a");
+    link.href = zipUrl;
+    link.download = `${selectedPlugin.title}.zip`;
+    link.click();
+    URL.revokeObjectURL(zipUrl);
+  }, [selectedPlugin]);
```
Similar to code-rabbit, lets add some error handling

devshun replied:
fixed this commit 

https://github.com/reearth/reearth-visualizer/pull/1216/commits/e2b6480d793f1b097fb55db1cd75aa676b14f23b

and added user FB when error occured

mulengawilfred commented on web/src/beta/features/PluginPlayground/Plugins/hook.ts:
```diff
@@ -0,0 +1,187 @@
+  };
+};
+
+const generateUniqueId = () => {
+  return (
+    Math.random().toString(36).substring(2, 15) +
+    Math.random().toString(36).substring(2, 15)
+  );
+};
```
`uuid` package is in the repo already so we can make use of it

mulengawilfred reviewed (changes requested):
Looks good. A few comments and some from code-rabbit that could be addressed.

devshun commented on web/src/beta/features/PluginPlayground/Plugins/hook.ts:
```diff
@@ -0,0 +1,188 @@
+      );
+    },
+    [selectedPlugin]
+  );
```
fix this commit 
https://github.com/reearth/reearth-visualizer/pull/1216/commits/ff7229a67425d4c82b0bc0b68663d987b411255b

devshun commented:
@mulengawilfred 

Thank you for your review. I have addressed the points you mentioned as well as the suggestions from CodeRabbit. Could you please review it at your convenience?🙏

devshun commented on web/src/beta/features/PluginPlayground/Plugins/utils.ts:
```diff
@@ -0,0 +1,38 @@
+      success: false,
+      message: "Invalid file type"
+    } as const;
+  }
```
thanks!
fixed this commit 

https://github.com/reearth/reearth-visualizer/pull/1216/commits/90f337c2c5d5eb33c97e726d812644f2a69ebb72
